dhall/src/Dhall/Core.hs

```diff
@@ -2269,3 +2238,58 @@ prettyURIComponent text =
 throws :: (Exception e, MonadIO io) => Either e a -> io a
 throws (Left  e) = liftIO (Control.Exception.throwIO e)
 throws (Right r) = return r
+
+{-
+Instead of converting explicitly between 'Expr's and 'MultiLet', it might
+be nicer to use a pattern synonym:
+
+> pattern MultiLet' :: NonEmpty (Binding s a) -> Expr s a -> Expr s a
+> pattern MultiLet' as b <- (multiLetFromExpr -> Just (MultiLet as b)) where
+>   MultiLet' as b = wrapInLets as b
+>
+> multiLetFromExpr :: Expr s a -> Maybe (MultiLet s a)
+> multiLetFromExpr = \case
+>     Let x mA a b -> Just (multiLet x mA a b)
+>     _ -> Nothing
+
+This works in principle, but GHC as of v8.8.1 doesn't handle it well:
+https://gitlab.haskell.org/ghc/ghc/issues/17096
+
+This should be fixed by GHC-8.10, so it might be worth revisiting then.
+-}
+
+{-| Generate a 'MultiLet' from the contents of a 'Let'.
+
+    In the resulting @'MultiLet' bs e@, @e@ is guaranteed not to be a 'Let',
+    but it might be a @('Note' … ('Let' …))@.
+
+    Given parser output, 'multiLet' consolidates @let@s that formed a
+    let-block in the original source.
+-}
+multiLet :: Text -> Maybe (Expr s a) -> Expr s a -> Expr s a -> MultiLet s a
+multiLet x0 mA0 a0 = \case
+    Let x1 mA1 a1 b1 ->
+        let MultiLet bs e = multiLet x1 mA1 a1 b1
+        in MultiLet (Data.List.NonEmpty.cons binding bs) e
+    e -> MultiLet (binding :| []) e
+  where
+    binding = Binding x0 mA0 a0
+
+{-| Wrap let-'Binding's around an 'Expr'. -}
+wrapInLets :: Foldable f => f (Binding s a) -> Expr s a -> Expr s a
```

Comment:
It seems like it would be more elegant if the type of `wrapInLets` were:

```haskell
wrapInLets :: MultiLet s a -> Expr s a
```

... but I don't feel strongly about this

Reply:
I have documented `wrapInLets`' relationship to `multiLet` in https://github.com/dhall-lang/dhall-haskell/pull/1242/commits/056da33fb204f10e14acc919cc2c6bee3d0e8552.

The advantage of the current type is that we can also use `wrapInLets` where we only have a `[Binding s a]`, not a `(NonEmpty (Binding s a))`.

Reply:
Alright, that sounds good